src/main/java/net/imglib2/algorithm/localextrema/LocalExtrema.java

```diff
@@ -92,79 +100,370 @@
 
 	/**
 	 * Find pixels that are extrema in their local neighborhood. The specific
-	 * test for being an extremum can is specified as an implementation of the
+	 * test for being an extremum can be specified as an implementation of the
 	 * {@link LocalNeighborhoodCheck} interface.
-	 * 
-	 * TODO: Make neighborhood shape configurable. This will require that Shape
-	 * can give a bounding box.
-	 * 
-	 * @param img
+	 *
+	 * The task is parallelized along the longest dimension of
+	 * <code>source</code> after adjusting for size based on <code>shape</code>.
+	 *
+	 * The number of tasks for parallelization is determined as:
+	 * <code>Math.max( Math.min( maxSizeDim, numThreads * 20 ), 1 )</code>
+	 *
+	 * where <code>maxSizeDim</code> is the longest dimension of
+	 * <code>img</code> after adjusting for the bounding box of a
+	 * {@link RectangleShape} with span 1, and numThreads is
+	 * <code>Runtime.getRuntime().availableProcessors()</code>
+	 *
+	 * {@link RectangleShape} is used as local neighborhood.
+	 *
+	 * Note: Pixels within 1 point of the <code>source</code> border will be
+	 * ignored as local extrema candidates because the complete neighborhood
+	 * would not be included in <code>source</code>. To include those pixel,
+	 * expand <code>source</code> accordingly. The returned coordinate list is
+	 * valid for the original <code>source</code>.
+	 *
+	 * @param source
+	 *            Find local extrema within this
+	 *            {@link RandomAccessibleInterval}
 	 * @param localNeighborhoodCheck
+	 *            Check if current pixel qualifies as local maximum.
 	 * @param service
-	 * @return
+	 *            {@link ExecutorService} handles parallel tasks
+	 * @return {@link ArrayList} of extrema
 	 */
-	public static < P, T extends Comparable< T > > ArrayList< P > findLocalExtrema( final RandomAccessibleInterval< T > img, final LocalNeighborhoodCheck< P, T > localNeighborhoodCheck, final ExecutorService service )
+	public static < P, T > ArrayList< P > findLocalExtrema( final RandomAccessibleInterval< T > source, final LocalNeighborhoodCheck< P, T > localNeighborhoodCheck, final ExecutorService service )
 	{
-		final ArrayList< P > allExtrema = new ArrayList< P >();
-
-		final Interval full = Intervals.expand( img, -1 );
-		final int n = img.numDimensions();
-		final int splitd = n - 1;
-		// FIXME is there a better way to determine number of threads
+		final RectangleShape shape = new RectangleShape( 1, true );
+		final long[] borderSize = getRequiredBorderSize( shape, source.numDimensions() );
+		final int nDim = source.numDimensions();
+		// Get biggest dimension after border subtraction. Parallelize along
+		// this dimension.
```

Comment:
I'm not sure this is always a good choice. Assuming some kind of flattened array, where the highest dimension is the slowest changing, it is probably preferable to choose that one because of cache utilisation (even if the extent is larger along another dimension). It would be interesting to see a benchmark.

In the end, of course it can be a rotated view or the highest dimension may have extent 1, etc...

Can you add a version where the parallelisation dimension is a parameter?